src/ql/pmgr/pass_types/base.cc

```diff
@@ -1120,68 +1125,79 @@ void Base::compile(
+            case NodeType::GROUP_IF: {
                 auto retval = run_main_pass(ir, context);
-                if (!condition->evaluate(retval)) {
-                    QL_IOUT("pass condition returned false, exiting loop");
-                    break;
+                if (condition->evaluate(retval)) {
+                    QL_IOUT("pass condition returned true, running sub-passes...");
+                    run_sub_passes(ir, context);
                 } else {
-                    QL_IOUT("pass condition returned true, continuing loop...");
+                    QL_IOUT("pass condition returned false, skipping " << sub_pass_order.size() << " sub-pass(es)");
                 }
-                run_sub_passes(ir, context);
+                break;
             }
-            break;
-        }
 
-        case NodeType::GROUP_REPEAT_UNTIL_NOT: {
-            QL_IOUT("entering loop pass loop...");
-            while (true) {
-                run_sub_passes(ir, context);
-                auto retval = run_main_pass(ir, context);
-                if (!condition->evaluate(retval)) {
-                    QL_IOUT("pass condition returned false, exiting loop");
-                    break;
-                } else {
-                    QL_IOUT("pass condition returned true, continuing loop...");
+            case NodeType::GROUP_WHILE: {
+                QL_IOUT("entering loop pass loop...");
+                while (true) {
+                    auto retval = run_main_pass(ir, context);
+                    if (!condition->evaluate(retval)) {
+                        QL_IOUT("pass condition returned false, exiting loop");
+                        break;
+                    } else {
+                        QL_IOUT("pass condition returned true, continuing loop...");
+                    }
+                    run_sub_passes(ir, context);
                 }
+                break;
             }
-            break;
-        }
 
-        default: QL_ASSERT(false);
-    }
+            case NodeType::GROUP_REPEAT_UNTIL_NOT: {
+                QL_IOUT("entering loop pass loop...");
+                while (true) {
+                    run_sub_passes(ir, context);
+                    auto retval = run_main_pass(ir, context);
+                    if (!condition->evaluate(retval)) {
+                        QL_IOUT("pass condition returned false, exiting loop");
+                        break;
+                    } else {
+                        QL_IOUT("pass condition returned true, continuing loop...");
+                    }
+                }
+                break;
+            }
+
+            default: QL_ASSERT(false);
+        }
 
-    // Handle configured debugging actions after running the pass.
-    handle_debugging(ir, context, true);
+        // Handle configured debugging actions after running the pass.
+        compile_phase = "debugging.after";
+        handle_debugging(ir, context, true);
 
+    } catch (utils::Exception &e) {
```

Comment:
Where do the exceptions that you want to catch come from? Only from handle_debugging or from executing the passes?
Or, why is this compile_phase relevant?
Or, why did you add this exception handler?

Reply:
Exceptions typically originate from QL_ICE and friends. This exception handler adds the pass name and 'phase' as context, and then rethrows. It was added because it was typically difficult to pinpoint where and when things went wrong, especially if an error occurred in code that's common between passes (such as `handle_debugging()`)